dango/lending/src/execute.rs

```diff
@@ -0,0 +1,177 @@
+        "Only the owner can whitelist denoms"
+    );
+
+    for (denom, _updates) in updates {
+        MARKETS.save(ctx.storage, &denom, &Market {})?;
+    }
+
+    Ok(Response::new())
+}
+
+/// Ensures that the sender's account is not a margin account.
+fn ensure_sender_account_is_not_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        !ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Margin accounts can't deposit or withdraw"
+    );
+    Ok(())
+}
+
+/// Ensures that the sender's account is a margin account.
+fn ensure_sender_account_is_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Only margin accounts can borrow and repay"
+    );
+    Ok(())
+}
+
+pub fn deposit(ctx: MutableCtx) -> anyhow::Result<Response> {
+    // Ensure margin accounts can't deposit
+    ensure_sender_account_is_not_margin(&ctx)?;
+
+    // For each deposited denom, ensure it's whitelisted and mint LP tokens.
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    for coin in ctx.funds {
+        ensure!(MARKETS.has(ctx.storage, &coin.denom), "Invalid denom");
+
+        let mut parts = vec![Part::from_str(NAMESPACE)?, Part::from_str("lp")?];
+        parts.extend_from_slice(coin.denom.inner());
+
+        let lp_denom = Denom::from_parts(parts)?;
+
+        msgs.push(Message::execute(
+            cfg.bank,
+            &bank::ExecuteMsg::Mint {
+                to: ctx.sender,
+                denom: lp_denom,
+                amount: coin.amount,
+            },
+            Coins::new(),
+        )?);
+    }
+
+    Ok(Response::new().add_messages(msgs))
+}
+
+pub fn withdraw(ctx: MutableCtx) -> anyhow::Result<Response> {
+    // Ensure margin accounts can't withdraw
+    ensure_sender_account_is_not_margin(&ctx)?;
+
+    // Ensure there are funds to withdraw
+    ensure!(!ctx.funds.is_empty(), "No funds to withdraw");
+
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    let mut withdrawn = Coins::new();
+    for coin in ctx.funds.into_iter() {
+        // Ensure only LP tokens are sent
+        let mut iter = coin.denom.inner().iter();
+
+        ensure!(
+            iter.next().map(|part| part.as_ref()) == Some(NAMESPACE),
+            "namespace:{NAMESPACE} not found"
+        );
+
+        ensure!(
+            iter.next().map(|part| part.as_ref()) == Some("lp"),
+            "namespace: lp not found"
+        );
```

Comment:
We can add a `Denom::starts_with` method to simplify this.

Also I don't think the `lp` part is necessary. `lending/...` is good enough.

Reply:
Fair, I just figured might as well add the lp part in case we ever need to mint another token from the lending namespace.